opencog/distvalue/CoverTree.cc

```diff
@@ -0,0 +1,576 @@
+{
+	if (pos != other.pos)
+		return false;
+
+	if (value != other.value)
+		return false;
+
+	for (auto elem : children)
+	{
+		auto res = std::find(other.children.begin(),other.children.end(), elem);
+
+		if (res == other.children.end())
+			return false;
+
+		if (elem != *res)
+			return false;
+	}
+	return true;
+}
+
+//CoverTree
+
+template <typename val_t>
+CoverTree<val_t>::CoverTree()
+	: _root_idx(-1) , _root_level(0) , _total_count(0) , _dims(1) {}
+
+template <typename val_t>
+CoverTree<val_t>::CoverTree(int dims)
+	:  _root_idx(-1) , _root_level(0) , _total_count(0) , _dims(dims) {}
+
+template <typename val_t>
+CoverTree<val_t>::CoverTree(CoverTreeNode<val_t> n,int dims)
+	:  _root_idx(-1) , _root_level(0) , _dims(dims) {
+	_nodes.push_back(n);
+	_root_idx = _nodes.size() - 1;
+    _total_count = get_count(n.value);
+}
+
+template <typename val_t>
+double CoverTree<val_t>::dist(const CoverTreeNode<val_t> & n1,
+                              const CoverTreeNode<val_t> & n2)
+{
+	return ::dist(n1.pos,n2.pos);
+}
+
+template <typename val_t>
+double CoverTree<val_t>::covdist(int level)
+{
+	return pow(2,level);
+}
+
+//TODO: Cache This?
+template <typename val_t>
+double CoverTree<val_t>::maxdist(const CoverTreeNode<val_t> & n) const
+{
+	double max = 0;
+	for (int c : n.children)
+	{
+		double tmp = dist(n,_nodes[c]);
+		if (tmp > max)
+			max = tmp;
+	}
+	return max;
+}
+
+template <typename val_t>
+int CoverTree<val_t>::popLeaf(CoverTreeNode<val_t> & n)
+{
+	int c_id = n.children.back();
+	CoverTreeNode<val_t> & c = _nodes[c_id];
+	if (c.children.size() == 0)
+	{
+		n.children.pop_back();
+		return c_id;
+	}
+	return popLeaf(c);
+}
+
+template <typename val_t>
+const CoverTreeNode<val_t> *
+CoverTree<val_t>::findNearestNeighbor(const DVec & pos) const
+{
+	const CoverTreeNode<val_t> & x = CoverTreeNode<val_t>(pos,val_t());
+	return findNearestNeighbor_(x,_nodes[_root_idx],&_nodes[_root_idx]);
+}
+
+template <typename val_t>
+const CoverTreeNode<val_t> *
+CoverTree<val_t>::findNearestNeighbor(const CoverTreeNode<val_t> & x) const
+{
+	return findNearestNeighbor_(x,_nodes[_root_idx],&_nodes[_root_idx]);
+}
+
+
+/* findNearestNeighbor implementation
+ * x is the node we want to find a neighbor for
+ * p is the node in the tree we are currently checking if it is a good neighbor
+ * y is the current nearest neighbor that we have found thus far
+ */
+template <typename val_t>
+const CoverTreeNode<val_t> *
+CoverTree<val_t>::findNearestNeighbor_(const CoverTreeNode<val_t> & x,
+									   const CoverTreeNode<val_t> & p,
+									   const CoverTreeNode<val_t> * y) const
+{
+	// If p is closer to x then y, update y
+	if (dist(x,p) < dist(x,*y))
+		y = &p;
+
+	// Create a Vector of indexs to the childref of p
+	// and sort it based on the distance of the child to x
+	std::vector<int> idxs(p.children.size());
+	std::size_t n(0);
+    std::generate(std::begin(idxs), std::end(idxs), [&]{ return n++; });
```

Comment:
Indentation problem.

Reply:
BTW, you can use iota, https://en.cppreference.com/w/cpp/algorithm/iota or https://www.boost.org/doc/libs/1_46_0/libs/range/doc/html/range/reference/algorithms/new/iota.html.